### PR TITLE
chore(HACBS-1717) New script to setup quay secret

### PR DESCRIPTION
**Pull request by scoheb**
Signed-off-by: Scott Hebert <shebert@redhat.com>

% ./setup-quay-push-secret.sh

```
Please enter the target workspace: [managed-release-team-tenant] temp4
Please enter the service account name: [release-service-account] 
Please enter the quay repo URL: [https://quay.io/repository/hacbs-release-tests/shebert-dc-metrocmap] 
Please enter your quay robot username: [] scoheb
Please enter your quay robot token: [] 
Going to create SPIAccessTokenBinding in temp4
spiaccesstokenbinding.appstudio.redhat.com/binding-dockerconfigjson-176186 created
Binding created
Waiting for AwaitingTokenData phase
spiaccesstokenbinding.appstudio.redhat.com/binding-dockerconfigjson-176186 condition met
Upload url: https://spi-oauth-spi-system.apps.scott-test-1717.clusters.stonesoupengineering.com/token/temp4/generated-spi-access-token-rr967
Waiting for Injected phase
spiaccesstokenbinding.appstudio.redhat.com/binding-dockerconfigjson-176186 condition met
Linked secret: robot-account-pull-secret-176186
```

% oc get sa/release-service-account -n temp3

```
NAME                      SECRETS   AGE
release-service-account   2         5m40s
```

% oc get sa/release-service-account -n temp3 -o yaml

```
apiVersion: v1
imagePullSecrets:
- name: release-service-account-dockercfg-j2c8g
kind: ServiceAccount
metadata:
  annotations:
    spi.appstudio.redhat.com/linked-access-token-binding: binding-dockerconfigjson-176186
  creationTimestamp: "2023-02-10T03:35:58Z"
  labels:
    spi.appstudio.redhat.com/managed-by-binding: binding-dockerconfigjson-176186
  name: release-service-account
  namespace: temp3
  resourceVersion: "13946611"
  uid: f689269b-05f2-4f30-8583-8a48ef89052d
secrets:
- name: release-service-account-dockercfg-j2c8g
- name: robot-account-pull-secret-176186
```

### CODE REVIEW

File: setup-quay-push-secret.sh [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+read -p "Please enter your quay robot username: [] " quay_oauth_user
+
+read -s -p "Please enter your quay robot token: [] " quay_oauth_token
+
+export binding_name=binding-dockerconfigjson-$$
+export robot_secret_name=robot-account-pull-secret-$$
+
+echo ""
+echo "Going to create SPIAccessTokenBinding in "$TARGET_NAMESPACE
```
Comment (davidmogar) [inline setup-quay-push-secret.sh:28]: Why not `echo "... in ${TARGET_NAMESPACE}". Cleaner, and variables should always be quoted.

File: setup-quay-push-secret.sh [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  repoUrl: $QUAY_REPO_URL
+  secret:
+    type: kubernetes.io/dockerconfigjson
+    name: ${robot_secret_name}
+    linkedTo:
+        - serviceAccount:
+            managed:
+              name: ${SERVICE_ACCOUNT}
+EOF
+echo 'Binding created'
```
Comment (davidmogar) [inline setup-quay-push-secret.sh:51]: Double quote it like in the other usages.

File: setup-quay-push-secret.sh [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+sleep 3
+echo "Waiting for AwaitingTokenData phase"
+kubectl wait  --for=jsonpath='{.status.phase}'=AwaitingTokenData  Spiaccesstokenbinding/${binding_name} -n $TARGET_NAMESPACE  --timeout=60s
+#kubectl get spiaccesstokenbinding/${binding_name} -n $TARGET_NAMESPACE -o  json | jq .
+
+UPLOAD_URL=$(kubectl get spiaccesstokenbinding/${binding_name} -n $TARGET_NAMESPACE -o  json | jq -r .status.uploadUrl)
+echo 'Upload url: '$UPLOAD_URL
```
Comment (davidmogar) [inline setup-quay-push-secret.sh:58]: Quote variables like I mentioned in my previous comment.

File: setup-quay-push-secret.sh [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+curl --insecure \
+  -H 'Content-Type: application/json' \
+  -H "Authorization: bearer $TOKEN" \
+  -d "{ \"access_token\": \"${quay_oauth_token}\" ,  \"username\": \"${quay_oauth_user}\" }" \
+  $UPLOAD_URL
+echo "Waiting for Injected phase"
+kubectl wait  --for=jsonpath='{.status.phase}'=Injected  Spiaccesstokenbinding/${binding_name} -n $TARGET_NAMESPACE  --timeout=60s
+
+LINKED_SECRET_NAME=$(kubectl get spiaccesstokenbinding/${binding_name} -n  ${TARGET_NAMESPACE}   -o  json | jq -r  .status.syncedObjectRef.name)
+echo 'Linked secret: '${LINKED_SECRET_NAME}
```
Comment (davidmogar) [inline setup-quay-push-secret.sh:68]: Quote variables like I mentioned in my previous comment.

File: setup-quay-push-secret.sh [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    name: ${robot_secret_name}
+    linkedTo:
+        - serviceAccount:
+            managed:
+              name: ${SERVICE_ACCOUNT}
+EOF
+echo 'Binding created'
+sleep 3
+echo "Waiting for AwaitingTokenData phase"
+kubectl wait  --for=jsonpath='{.status.phase}'=AwaitingTokenData  Spiaccesstokenbinding/${binding_name} -n $TARGET_NAMESPACE  --timeout=60s
```
Comment (davidmogar) [inline setup-quay-push-secret.sh:54]:
```suggestion
kubectl wait --for=jsonpath='{.status.phase}'=AwaitingTokenData Spiaccesstokenbinding/${binding_name} -n $TARGET_NAMESPACE --timeout=60s
```

File: setup-quay-push-secret.sh [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+#kubectl get spiaccesstokenbinding/${binding_name} -n $TARGET_NAMESPACE -o  json | jq .
+
+UPLOAD_URL=$(kubectl get spiaccesstokenbinding/${binding_name} -n $TARGET_NAMESPACE -o  json | jq -r .status.uploadUrl)
```
Comment (davidmogar) [inline setup-quay-push-secret.sh:57]:
```suggestion
UPLOAD_URL=$(kubectl get spiaccesstokenbinding/${binding_name} -n $TARGET_NAMESPACE -o json | jq -r .status.uploadUrl)
```

Comment (davidmogar) [reply setup-quay-push-secret.sh:57]: If feels weird to me that you use capital letters for these variables. I reserve those for constants on the top of the file.

File: setup-quay-push-secret.sh [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+echo 'Upload url: '$UPLOAD_URL
+curl --insecure \
+  -H 'Content-Type: application/json' \
+  -H "Authorization: bearer $TOKEN" \
+  -d "{ \"access_token\": \"${quay_oauth_token}\" ,  \"username\": \"${quay_oauth_user}\" }" \
+  $UPLOAD_URL
+echo "Waiting for Injected phase"
+kubectl wait  --for=jsonpath='{.status.phase}'=Injected  Spiaccesstokenbinding/${binding_name} -n $TARGET_NAMESPACE  --timeout=60s
```
Comment (davidmogar) [inline setup-quay-push-secret.sh:65]:
```suggestion
kubectl wait --for=jsonpath='{.status.phase}'=Injected Spiaccesstokenbinding/${binding_name} -n $TARGET_NAMESPACE --timeout=60s
```

File: setup-quay-push-secret.sh [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+LINKED_SECRET_NAME=$(kubectl get spiaccesstokenbinding/${binding_name} -n  ${TARGET_NAMESPACE}   -o  json | jq -r  .status.syncedObjectRef.name)
```
Comment (davidmogar) [inline setup-quay-push-secret.sh:67]:
```suggestion
LINKED_SECRET_NAME=$(kubectl get spiaccesstokenbinding/${binding_name} -n ${TARGET_NAMESPACE} -o json | jq -r .status.syncedObjectRef.name)
```

File: setup-quay-push-secret.sh [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+      - type: rw
+        area: registry
+      - type: rw
+        area: registryMetadata
+  repoUrl: $QUAY_REPO_URL
+  secret:
+    type: kubernetes.io/dockerconfigjson
+    name: ${robot_secret_name}
+    linkedTo:
+        - serviceAccount:
```
Comment (davidmogar) [inline setup-quay-push-secret.sh:47]: Spacing is not consistent. This is using 4 white spaces while other lines use 2.

File: setup-quay-push-secret.sh (hunk 1 above)
Comment (scoheb) [reply setup-quay-push-secret.sh:28]: fixed

File: setup-quay-push-secret.sh [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+            managed:
+              name: ${SERVICE_ACCOUNT}
+EOF
+echo 'Binding created'
+sleep 3
+echo "Waiting for AwaitingTokenData phase"
+kubectl wait  --for=jsonpath='{.status.phase}'=AwaitingTokenData  Spiaccesstokenbinding/${binding_name} -n $TARGET_NAMESPACE  --timeout=60s
+#kubectl get spiaccesstokenbinding/${binding_name} -n $TARGET_NAMESPACE -o  json | jq .
```
[the remark this answers is not in the document]
Comment (scoheb) [reply setup-quay-push-secret.sh:55]: fixed

File: setup-quay-push-secret.sh (hunk 2 above)
Comment (scoheb) [reply setup-quay-push-secret.sh:51]: fixed

File: setup-quay-push-secret.sh (hunk 4 above)
Comment (scoheb) [reply setup-quay-push-secret.sh:68]: fixed

File: setup-quay-push-secret.sh (hunk 5 above)
Comment (scoheb) [reply setup-quay-push-secret.sh:54]: fixed

File: setup-quay-push-secret.sh [hunk 11]
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
+
+# This script collects input needed to create a dockerconfigjson secret
+# and link it to a newly created service account
+
+export TARGET_NAMESPACE_DEFAULT=managed-release-team-tenant
+export QUAY_REPO_URL_EXAMPLE=https://quay.io/repository/hacbs-release-tests/shebert-dc-metrocmap
```
[the remark this answers is not in the document]
Comment (scoheb) [reply setup-quay-push-secret.sh]: Yes. This is something new. skopeo/quay does not like a user/pass combo in the dockerconfig file. This uses SPI to generate the auth token in the dockerconfig file.

File: setup-quay-push-secret.sh (hunk 7 above)
Comment (scoheb) [reply setup-quay-push-secret.sh]: fixed

File: setup-quay-push-secret.sh (hunk 8 above)
Comment (scoheb) [reply setup-quay-push-secret.sh]: fixed

File: setup-quay-push-secret.sh (hunk 9 above)
Comment (scoheb) [reply setup-quay-push-secret.sh]: fixed

File: setup-quay-push-secret.sh (hunk 6 above)
Comment (scoheb) [reply setup-quay-push-secret.sh]: re-worked all variables.

**Comment by scoheb** [comment]
fixed all comments and made some improvements to variables and error handling

**Review by johnbieren — APPROVED** [review]

**Pull request merged**